internal/catalogmetadata/types.go

```diff
@@ -140,6 +145,10 @@ func (b *Bundle) propertiesByType(propType string) []*property.Property {
 	return b.propertiesMap[propType]
 }
 
+func (b *Bundle) HasDeprecation() bool {
```

Comment:
Nit-picky but maybe:
```suggestion
func (b *Bundle) IsDeprecated() bool {
```
reads a little better to me.

Reply:
I can see that. Reason I went with `HasDeprecation` is more so because of the potentially confusing classification of "Deprecated" in this case. If a channel is deprecated, that doesn't necessarily mean _every_ bundle in that channel is deprecated. The `ClusterExtension` would be marked deprecated if it specifically asks for that channel but otherwise that deprecation notice wouldn't show up. Due to this I felt `HasDeprecation` was more like asking "Are there any deprecations that are associated with this bundle?" rather than "Is _this_ bundle deprecated?"

That being said, I'm not opposed to changing this. I do feel that `IsDeprecated()` could be a bit misleading without having a deeper understanding of how the deprecation concept works.

Reply:
Ah good point, and as you noted I myself also misunderstood the concept here by thinking that all bundles under a deprecated channel were also considered deprecated. If that's not the case then I agree with `HasDeprecation()`.

Reply:
Sounds like a good place to add GoDoc :)

Reply:
https://github.com/operator-framework/operator-controller/pull/574/commits/a2ff3ff37ae6f532db7c425aa4798eed57537f06

Reply:
One thing to point out with this that I thought about while writing the GoDoc is that during resolution if the bundle is associated with _any_ deprecations it will be less preferred over other bundles even if it is an olm.channel deprecation that isn't being considered due to the specification of using a different channel.

For example, given the following set of bundles in a desired channel `stable`:
```
foo.v1.0.0
foo.v1.1.0
foo.v2.0.0
foo.v2.1.0
foo.v3.0.0 <-- associated with a deprecated channel named stable-v3 due to $reasons
```

The resulting resolution list in order of most preferred -> least preferred would be:
```
foo.v2.1.0
foo.v2.0.0
foo.v1.1.0
foo.v1.0.0
foo.v3.0.0
```

This feels like a scenario we may want to avoid by having a distinction between a bundle having explicit deprecation due to the package or itself being deprecated and the looser deprecation associated with the bundle being an entry within a deprecated channel

Reply:
cc @joelanford @grokspawn ^

Reply:
I've updated the implementation in https://github.com/operator-framework/operator-controller/pull/574/commits/9186d16a4353a45131e3e460cf6ffa9d5d3886f9 to avoid the scenario I mentioned above